src/pip/_internal/vcs/git.py

```diff
@@ -362,8 +362,8 @@ def update_submodules(cls, location):
 
     @classmethod
     def controls_location(cls, location):
-        if super(Git, cls).controls_location(location):
-            return True
+        if not super(Git, cls).controls_location(location):
```

Comment:
@chrahunt: Thanks for taking the time to review.
My assumption was that the .git/.hg/.svn/.bzr directory would always be present in the top level repository directory and we could rely on that to detect vcs type quickly.  This assumption appears to be incorrect.

The subprocess call to the vcs tool takes some time. Calling `hg` appears to be a little slower than calling `git` (on my widows machine).
`git rev-parse` takes around 60mS
`hg identify` takes around 130mS

Currently the method used by `VcsSupport.get_backend_for_dir()` is:
- Iterate over each vcs type:
  - Look for the appropriate .git/.hg/.svn/.bzr directory for the vcs type in this iteration
  - If not found then allow the vcs tool to be called to perform the detection.  Prior to this PR only `git` implemented this call.  This PR adds a `hg` call. 

I'm thinking it would be preferable to avoid (where possible) calling each vcs tool until the correct one is found.  Rather using directory inspection for ALL vcs types first and falling back to calling vcs tools if inspection finds nothing.

With this in mind, would the following be a better approach for `VcsSupport.get_backend_for_dir()`?
- Search every directory up the directory tree for a .git/.hg/.svn/.bzr directory.
  - if one is found then assume that is the vcs type, i.e if `.git` directory is found then vcs is assumed to be `git` without calling the vcs tool.
- If no .git/.hg/.svn/.bzr directory is found, then call each vcs tool that implements `controls_location()` until one reports that it controls the location.

Perhaps this optimization belongs in a separate PR?

Reply:
Yes, in the interest of landing the functional changes (maybe even in time for 19.3) I think a separate PR for the optimization would be appropriate. If you want to create an issue for it now then it could accumulate feedback in the meantime. The approach seems reasonable, but I'll save my feedback for the issue so it doesn't distract from your changes.

Reply:
Have raised #7162 to capture the optimization.

**Note:** Reverting back to the original behavior requires changes to `run_command()` and `call_subprocess()` to allow a vcs call to fail without logging errors.

Reply:
I think a general refactoring of `call_subprocess` would be in order, to return an object that has exit code, output, etc, but it is out of scope of this PR. I have made a comment [here](https://github.com/pypa/pip/pull/7072#discussion_r333800125) on the new code, so I will resolve this conversation.